music21/musicxml/m21ToXml.py

```diff
@@ -2676,17 +2720,22 @@ def getXmlScorePart(self):
 
         # TODO: part-abbreviation-display
         # TODO: group
-
-        # TODO: unbounded...
-        i = self.firstInstrumentObject
-
-        if (i.instrumentName is not None or i.instrumentAbbreviation is not None
-                or i.midiProgram is not None):
-            mxScorePart.append(self.instrumentToXmlScoreInstrument(i))
-
-        # TODO: midi-device
-        if i.midiProgram is not None:
-            mxScorePart.append(self.instrumentToXmlMidiInstrument(i))
+        for inst in self.instrumentStream:
+            # only use the first instance of this class
+            if inst is not self.instrumentStream[type(inst)].first():
+                continue
+            if (inst.instrumentName is not None
+                    or inst.instrumentAbbreviation is not None
+                    or inst.midiProgram is not None):
+                mxScorePart.append(self.instrumentToXmlScoreInstrument(inst))
+
+        for inst in self.instrumentStream:
```

Comment:
this looks like it's going through the stream twice.  Is there another way of making it clear that we're doing different code for the first instrument of each type in the stream, by iterating once over self.instrumentStream and getting the first of each one?

Reply:
Ah, I see there's ScoreInstrument and MidiInstrument -- perhaps a comment could draw attention to that.  Such as:

"The first instrument of each Class appears as a Score instrument."  And later "Subsequent instruments appear as MIDI instruments"

Reply:
think this is better now -- two loops purely for the sake of writing the tags in the right order. these are tiny streams of just a few elements. (unless you have instrument changes every single note, which we aren't optimizing for)